test/e2e/end_user_test.go

```diff
@@ -163,4 +128,72 @@ var _ = Describe("Openshift on Azure end user e2e tests [EndUser]", func() {
 		fmt.Println(result.Error().Error())
 		Expect(result.Error().Error()).To(ContainSubstring("pods \"sync-master-000000\" is forbidden: User \"enduser\" cannot get pods/log in the namespace \"kube-system\""))
 	})
+
+	It("should deploy a template with persistent storage and test failure modes", func() {
+		By(fmt.Sprintf("instantiating the template and getting the route (%v)", time.Now()))
+		route, err := c.instantiateTemplate("django-psql-persistent")
+		Expect(err).NotTo(HaveOccurred())
+		// make sure only 1 ingress point is returned
+		Expect(len(route.Status.Ingress)).To(Equal(1))
+		host := route.Status.Ingress[0].Host
+		url := fmt.Sprintf("http://%s", host)
+
+		// Curl the endpoint and search for a string
+		httpc := &http.Client{
+			Timeout: 10 * time.Second,
+		}
+		var regex = regexp.MustCompile(`Page views:\s*(\d+)`)
+		By(fmt.Sprintf("hitting the route %s 3 times, expecting counter to increment (%v)", url, time.Now()))
+		var prevCounter, currCounter int
+		for i := 0; i < 3; i++ {
+			resp, err := httpc.Get(url)
+			Expect(err).NotTo(HaveOccurred())
+			defer resp.Body.Close()
+			Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+			contents, err := ioutil.ReadAll(resp.Body)
+			Expect(err).NotTo(HaveOccurred())
+			matches := regex.FindStringSubmatch(string(contents))
+			Expect(matches).NotTo(BeNil())
+			currCounter, err = strconv.Atoi(matches[1])
+			Expect(err).NotTo(HaveOccurred())
+			Expect(currCounter).Should(BeNumerically(">", prevCounter))
+			prevCounter = currCounter
+			time.Sleep(time.Second)
+		}
+
+		// Find the postgresql pod and delete it so it gets re-created
+		By(fmt.Sprintf("searching for the database pod (%v)", time.Now()))
+		listOptions := metav1.ListOptions{
+			LabelSelector: "name=postgresql",
+		}
+		pods, err := c.kc.CoreV1().Pods(c.namespace).List(listOptions)
+		Expect(err).NotTo(HaveOccurred())
+		pgsqlPod := pods.Items[0]
+		By(fmt.Sprintf("deleting the database pod (%v)", time.Now()))
+		err = c.kc.CoreV1().Pods(c.namespace).Delete(pgsqlPod.Name, &metav1.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		// Pod should be re-created automatically
+		time.Sleep(5 * time.Second)
```

Comment:
You shouldn't be sleeping here and rely only on PollImmediate

Reply:
@kargakis it appears to take a bit for the pod to go from Ready -> Deleting so the test flakes some times because the polling returns true before the pod is deleted and the http GET failes later on

Reply:
deleting the pod and waiting an arbitrary amount of time is going to be flaky. A more robust solution would be to scale the deployment/statefulset to zero, wait for zero pods to exist, then scale back to one, wait for one pod to exist.